vmaas/reposcan/katello.py

```diff
@@ -0,0 +1,181 @@
+"""
+Module containing code interacting with Satellite/Katello server.
+"""
+import os
+import shutil
+import tempfile
+
+import requests
+
+from vmaas.common.logging_utils import get_logger
+from vmaas.reposcan.database.repository_store import RepositoryStore
+
+LOGGER = get_logger(__name__)
+
+KATELLO_CA_CERT_PATH = "/pub/katello-server-ca.crt"
+KATELLO_ACCESS_CERT_PATH = "/katello/api/v2/organizations/%s/download_debug_certificate"
+KATELLO_ORG_LIST_API = "/katello/api/v2/organizations"
+KATELLO_REPOS_LIST_API = "/katello/api/v2/repositories?organization_id=%s&search=redhat=true"
+
+RETRY_COUNT = int(os.getenv("RETRY_COUNT", "3"))
+
+
+class KatelloApi:
+    """Class used to obtain information from Satellite/Katello API."""
+    def __init__(self, hostname: str, api_user: str, api_pass: str):
+        self.hostname: str = hostname
+        self.api_user: str = api_user
+        self.api_pass: str = api_pass
+
+        self.tmp_directory: str | None = None
+
+        repo_store: RepositoryStore = RepositoryStore()
+        self.archs: list[str] = repo_store.list_archs()
+
+        self.success: bool = False
+
+    def _rm_tmpdir(self) -> None:
+        if self.tmp_directory:
+            LOGGER.debug("Deleting tmp dir: %s", self.tmp_directory)
+            shutil.rmtree(self.tmp_directory)
+        self.tmp_directory = None
+
+    def _find_arch(self, url: str) -> str | None:
+        for arch in self.archs:
+            if arch in url:
+                return arch
+        return None
+
+    def _fetch_katello(  # pylint: disable=too-many-branches
+        self, endpoint, scheme="https", json=True, timeout=30, **kwargs
+    ) -> dict:
+        """Sends request to Katello server/API"""
+
+        if scheme == "https":
+            verify = os.path.join(self.tmp_directory, os.path.basename(KATELLO_CA_CERT_PATH))
+            if not os.path.isfile(verify):
+                LOGGER.error("Katello CA certificate not found! Skipping API call.")
+                self.success = False
+                return {}
+            auth = (self.api_user, self.api_pass)
+        else:
+            verify = True
+            auth = None
+
+        if json:
+            headers = {"Accept": "application/json"}
+        else:
+            headers = {}
+
+        url = f"{scheme}://{self.hostname}{endpoint}"
+        tries = 0
+
+        while True:
+            if tries >= RETRY_COUNT:
```

Comment:
DRY

Reply:
it would be better to use something like https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries

but at least set some timeout between retries to not overload katello API